### PR TITLE
some stacks questions

### PR DESCRIPTION
**Pull request by manumewari**

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (sharadJay) [inline .idea/vcs.xml:1]: put it on gitgnore

File: stacks/exercise1 [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/* https://github.com/kuychaco/algoClass/blob/master/data-structures/stack.js */
```
Comment (sharadJay) [inline stacks/exercise1]: remove all comments and console logs

File: stacks/exercise1 [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    this.index = 0;
+}
+
+Stack.prototype.push = function(num) {
+    console.log("current size of stack is "+ this.index);
+    this.stackObj[ this.index++ ] = num;
+}
+
+Stack.prototype.pop = function() {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/exercise1]: why not use count function from line 54?

File: stacks/exercise1 [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        console.log("there are no elements in the stack");
+    }
+    else
+    {
+        console.log("this.index ==> "+this.index);
+        console.log("popping element "+ this.stackObj[this.index-1]);
+        this.stackObj[ this.index-- ] = '';
```
Comment (sharadJay) [inline stacks/exercise1]: this element should have been deleted instead of assigning blank value. Assigning blank would not make any difference.
Also, pop should be returning element that got deleted.

File: stacks/exercise1 [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+
+Stack.prototype.peek = function() {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/exercise1]: should be returning the last element

File: stacks/exercise1 [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        console.log("this.index ==> "+this.index);
+        console.log("last element is "+ this.stackObj[this.index-1]);
+    }
+}
+
+Stack.prototype.count = function() {
+    if( this.index === 0 )
+    {
+        console.log("there are no elements in the stack");
```
Comment (sharadJay) [inline stacks/exercise1]: would it be more correct if it just returned this.index?

File: stacks/stack.test.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    var testStack = new Stack();
+
+    expect(testStack.count()).toBe(0);
+
+});
+
+test('Count should decreasewhen element is popped from stack', () => {
+    var testStack = new Stack();
+
+    testStack.push(10);
```
Comment (sharadJay) [inline stacks/stack.test.js]: Be precise and sharp
```javascript
[10, 11, 12, 13].each(elem => testStack.push(elem))
```

File: stacks/stack.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
+
+function Stack(){
+    this.stackObj = {};
+    this.index = 0;
+    this.maxSize = 10;
```
Comment (sharadJay) [inline stacks/stack.js]: should not the max size be passed in constructor call with a default value?

File: stacks/stack.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        console.log("Max capacity already reached. Remove element before adding a new one.");
+    }
+    else
+    {
+        this.stackObj[ this.index++ ] = num;
+    }
+}
+
+Stack.prototype.pop = function() {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/stack.js:21]: no need for if statement, to get the console logs

File: stacks/stack.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        console.log("there are no elements in the stack");
+    }
+    else
+    {
+        return this.stackObj[this.index-1];
+    }
+}
+
+Stack.prototype.count = function() {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/stack.js]: what if this function only returned this.index? wouldn't it work the same?

File: stacks/stack.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+Stack.prototype.until = function(num) {
+    if( this.index === 0 )
+    {
+        console.log("there are no elements in the stack");
+    }
+    else
+    {
+        var popCount = 0;
```
Comment (sharadJay) [inline stacks/stack.js]: what if this utilized what you already have written
```javascript
let shouldContinuePoppingElements = true;
do {
   let poppedElement = this.pop();
   shouldContinuePoppingElements = !(poppedElement == num)
}
while(shouldContinuePoppingElements)
```

File: stacks/exercise4 [hunk 12]
```diff
@@ -0,0 +1,55 @@
+
+/* https://github.com/kuychaco/algoClass/blob/master/data-structures/stack.js */
+
+/*** Sort a stack so that its elements are in ascending order. */
+
+
+/* NOT COMPLETE */
+
+function Stack(){
```
Comment (sharadJay) [inline stacks/exercise4]: you should import stack from stack.js

File: stacks/exercise4 [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+Stack.prototype.min = function() {
+    if( this.index === 0 )
+    {
+        console.log("there are no elements in the stack");
+    }
+    else
+    {
+        console.log("this.index ==> "+this.index);
+        console.log("smallest element is "+ this.minObj[this.index-1]);
```
Comment (sharadJay) [inline stacks/exercise4]: how is this working? what is minObj?

File: stacks/stack.test.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+    testStack.push(12);
+    expect(testStack.count()).toBe(3);
+
+    testStack.pop();
+    expect(testStack.count()).toBe(2);
+});
+
+test('Count should be 0 when no element is pushed in stack', () => {
+    var testStack = new Stack();
```
Comment (sharadJay) [inline stacks/stack.test.js]: why not
```javascript
 expect(new Stack().count()).toBe(0);
```

File: stacks/stack.test.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    expect(testStack.count()).toBe(5);
+
+    testStack.push(15);
+    expect(testStack.count()).toBe(5);
+});
+
+test('peek() should return same element once stack is reached', () => {
+    var testStack = new Stack(5);
+
+[10, 11, 12, 13, 14].forEach(element => testStack.push(element));
```
Comment (sharadJay) [inline stacks/stack.test.js]: correct indentation.

File: stacks/stack.test.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+expect(testStack.peek()).toBe(14);
+
+testStack.push(15);
+expect(testStack.peek()).toBe(14);
```
Comment (sharadJay) [inline stacks/stack.test.js]: shouldn't test fail here?

File: stacks/stack.js [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+Stack.prototype.until = function(num) {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/stack.js]: why is this needed?

File: stacks/stack.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        console.log("there are no elements in the stack");
+    }
+    else
+    {
+        var deletedNumber = '';
```
Comment (sharadJay) [inline stacks/stack.js]: why declare it outside loop?

File: stacks/stack.js [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        return this.stackObj[this.index-1];
+    }
+}
+
+Stack.prototype.count = function() {
+    return this.index;
+}
+
+Stack.prototype.contains = function(num) {
+    if( this.index === 0 )
```
Comment (sharadJay) [inline stacks/stack.js]: remove if condition, no need to check this

File: stacks/minstack.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
+
+function MinStack(){
```
Comment (sharadJay) [inline stacks/minstack.js]: why not use existing implementation of stack like this https://github.com/sharadJay/datastructures-algorithms-javascript/blob/master/datastructures/Stacks/minStack.js

File: stacks/bracketstester.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
+const Stack = require('./Stack');
+
+function BracketsTester(stackMaxCapacity) {
+    this.stringStack = new Stack(stackMaxCapacity);
+    this.count = 0;
```
Comment (sharadJay) [inline stacks/bracketstester.js]: why do you need count in here? and why do you even need to create a class for this problem, this could just be a function checkBrackets

File: stacks/bracketstester.js [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+BracketsTester.prototype.checkBrackets = function(stringValue) {
+
+    Array.from(stringValue).forEach(charValue => this.stringStack.push(charValue));
+
+    while ( this.stringStack.peek() ) {
```
Comment (sharadJay) [inline stacks/bracketstester.js]: you are using falsy as check a check for going forward but what else can be falsy that you can get in your function
1. 0 or "0"
2. " " or ""
3. null or undefined
either the check should have been ``` typeof this.stringStack.peek() == undefined ``` or you could have used stack.count?

File: stacks/bracketstester.test.js [hunk 23]
```diff
@@ -0,0 +1,10 @@
+const BracketsTester = require('./BracketsTester');
+
+test('test ', () => {
```
Comment (sharadJay) [inline stacks/bracketstester.test.js]: write some more tests
1. empty string
2. argument as number
3. unbalanced brackets

File: stacks/bracketstester.js [hunk 24]
```diff
@@ -0,0 +1,30 @@
+const Stack = require('./Stack');
+
+function BracketsTester(stackMaxCapacity) {
+    this.stringStack = new Stack(stackMaxCapacity);
+    this.count = 0;
+}
+
+BracketsTester.prototype.checkBrackets = function(stringValue) {
+
+    Array.from(stringValue).forEach(charValue => this.stringStack.push(charValue));
```
Comment (sharadJay) [inline stacks/bracketstester.js]: what if you only pushed brackets in the stacks, would have saved time when you started verifying it.

File: stacks/bracketstester.js [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+
+    while ( this.stringStack.peek() ) {
+
+        var poppedElement = this.stringStack.pop();
+
+        if(this.count === 0 && poppedElement === "(") {//If opening braces comes first -> stop.
```
Comment (sharadJay) [inline stacks/bracketstester.js]: should not need this right because count would not be 0 when you return final result, isnt it?

File: stacks/minstack.js [hunk 26]
```diff
@@ -0,0 +1,39 @@
+const Stack = require('./Stack');
+
+function MinStack(stackMaxSize){
+    this.stack = new Stack(stackMaxSize);
+    this.minStack = new Stack(stackMaxSize);
+}
+
+MinStack.prototype.push = function(num) {
+    var currentMinimum = this.minStack.peek();
+    if(this.minStack.count() === 0 || (!currentMinimum || currentMinimum > num) ) //First element
```
Comment (sharadJay) [inline stacks/minstack.js]: welcome to javascript, let's learn about truthy and falsy values https://www.sitepoint.com/javascript-truthy-falsy/
!currentMinimum would be true if currentMinimum is 0

File: stacks/minstack.js [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+    this.stack.push(num);
+}
+
+MinStack.prototype.pop = function() {
+    this.minStack.pop();
+    this.stack.pop();
+}
+
+MinStack.prototype.min = function() {
+    if( this.minStack.count() === 0 )
```
Comment (sharadJay) [inline stacks/minstack.js]: dont add console logs, this would not work with your tests anyways.

File: stacks/minstack.js [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+    else
+    {
+        this.minStack.push(currentMinimum);
+    }
+    this.stack.push(num);
+}
+
+MinStack.prototype.pop = function() {
+    this.minStack.pop();
+    this.stack.pop();
```
Comment (sharadJay) [inline stacks/minstack.js]: this should be returning stack.pop right?

File: stacks/bracketstester.js (hunk 25 above)
Comment (manumewari) [reply stacks/bracketstester.js]: this is for some string like - "Math.min(5,(6-3))(3*4" if last occurrence of bracket is "(" in string then it will return false as it is not closed. first bracket found from stack should be ")".

File: stacks/minstack.test.js [hunk 29]
```diff
@@ -0,0 +1,41 @@
+const MinStack = require('./MinStack');
+
+test('min should return the smallest item in stack', () => {
```
Comment (sharadJay) [inline stacks/minstack.test.js:3]: you need tests on peek

File: stacks/bracketstester.js (hunk 21 above)
Comment (manumewari) [reply stacks/bracketstester.js]: i am using count for counting the number of brackets. When a ")" is found, count is increased. and when a opening bracket "(" is found, count is decreased. So in the end, if brackets are balanced, count would remain 0.

File: stacks/stack.js [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        if(this.stackObj[i] === num)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+Stack.prototype.until = function(num) {
+    for( var i = this.index-1 ; i>=0 ; i-- )
```
Comment (sharadJay) [inline stacks/stack.js]: will this work without any loops with while statement? 
```javascript
while(this.pop() === num || typeof this.peek() !== undefined)
```

**Review by sharadJay — CHANGES_REQUESTED** [review]
Needs more brevity and suspicion towards falsy values.

File: stacks/bracketstester.js (hunk 22 above)
Comment (manumewari) [reply stacks/bracketstester.js]: This will work fine as here it is checked as string and “0” is will not be considered as integer 0. 
“ “ is also not considered as empty string.
“” will be treated as false as expected.

A sample console log tested with string ("sqrt(5*(3+8)/ (4-0) )") with 0 and empty spaces “ “:
poppedElement : ), this.count : 0
poppedElement :  , this.count : 1
poppedElement : ), this.count : 1
poppedElement : 0, this.count : 2
poppedElement : -, this.count : 2
poppedElement : 4, this.count : 2
poppedElement : (, this.count : 2
poppedElement :  , this.count : 1
poppedElement : /, this.count : 1
poppedElement : ), this.count : 1
poppedElement : 8, this.count : 2
poppedElement : +, this.count : 2
poppedElement : 3, this.count : 2
poppedElement : (, this.count : 2
poppedElement : *, this.count : 1
poppedElement : 5, this.count : 1
poppedElement : (, this.count : 1
poppedElement : t, this.count : 0
poppedElement : r, this.count : 0
poppedElement : q, this.count : 0
poppedElement : s, this.count : 0

File: stacks/bracketstester.test.js (hunk 23 above)
Comment (manumewari) [reply stacks/bracketstester.test.js]: added more

File: stacks/bracketstester.js (hunk 24 above)
Comment (manumewari) [reply stacks/bracketstester.js]: good idea, now adding only brackets

File: stacks/minstack.js (hunk 26 above)
Comment (manumewari) [reply stacks/minstack.js]: i had assumed it to be for 1 or more number. missed a condition with 0 numbers. corrected now.

File: stacks/minstack.js (hunk 27 above)
Comment (manumewari) [reply stacks/minstack.js]: removed

File: stacks/minstack.js (hunk 28 above)
Comment (manumewari) [reply stacks/minstack.js]: here was not using the value of popped element. so did not return the value.

File: stacks/bracketstester.js (hunk 22 above)
Comment (sharadJay) [reply stacks/bracketstester.js]: mil toh gaya hi hoga falsy wala problem

File: stacks/bracketstester.js (hunk 25 above)
Comment (sharadJay) [reply stacks/bracketstester.js]: ha but jab pura program khatm bho jaega then wouldnt count be -1

File: stacks/minstack.js (hunk 28 above)
Comment (sharadJay) [reply stacks/minstack.js]: a pop always returns value that got deleted

File: stacks/checkbrackets.js [hunk 31]
```diff
@@ -0,0 +1,29 @@
+const Stack = require('./Stack');
+
+function checkBrackets(stringValue) {
```
Comment (sharadJay) [inline stacks/checkbrackets.js:3]: You should make this extensible in the sense that there should be array of both openingBrackets and closing brackets
const openingBrackets = ["(", "{", "["]
const closingBrackets = [")", "}", "]"]
as mathematical operation can have so much more.

File: stacks/checkbrackets.test.js [hunk 32]
```diff
@@ -0,0 +1,43 @@
+const checkBrackets = require('./checkBrackets');
+
+test('test if the brackets are balanced, each opening bracket has one closing bracket then checkBrackets() should return true', () => {
+    var testString = "sqrt(5*(3+8)/ (4-0) )";
+    var testStack = checkBrackets(testString);
+    expect(testStack).toBeTruthy();
+});
+
+test('test if the brackets are not balanced then checkBrackets() should return false', () => {
```
Comment (sharadJay) [inline stacks/checkbrackets.test.js:9]: two tests can not have same name line 9 and line 15

File: stacks/minstack.js [hunk 33]
```diff
@@ -0,0 +1,32 @@
+const Stack = require('./Stack');
+
+function MinStack(stackMaxSize){
+    this.stack = new Stack(stackMaxSize);
+    this.minStack = new Stack(stackMaxSize);
+}
+
+MinStack.prototype.push = function(num) {
+    var currentMinimum = this.minStack.peek();
+    if(this.minStack.count() === 0 || (typeof currentMinimum === "undefined" || currentMinimum > num) )
```
Comment (sharadJay) [inline stacks/minstack.js:11]: this needs more brevity, ideally, number to be pushed in minStack should be decided in if else loop and push in minstack should happen outside it.

**Review by sharadJay — CHANGES_REQUESTED** [review]
added more comments

File: stacks/bracketstester.js (hunk 22 above)
Comment (manumewari) [reply stacks/bracketstester.js]: as it is a string, there was no problem here as " " and "0" are also string object received.

File: stacks/bracketstester.js (hunk 25 above)
Comment (manumewari) [reply stacks/bracketstester.js]: there are 3 conditions:
1.) if last bracket occurrence is "(" which means its not closed then first if condition will return false without checking further. Ex. "Math.min(5,(6-3))**(**3*4"

Comment (manumewari) [reply stacks/bracketstester.js]: There are 3 conditions:
        if(count === 0 && poppedElement === "(") { return false; }
        else if ( poppedElement === ")" ) { count ++; }
        else if ( poppedElement === "(" ) { count --; }
1.) For balanced brackets, ")" will always come last. If last bracket occurrence is "(" which means its not closed then first if condition will return false without checking further. 
Ex. "5*(6-3)(3*4".
2.) If first condition is not the case, count will increment when ")" is found and decrease when ")" is found. So, when the program ends..... count will be zero if brackets are balanced.

File: stacks/minstack.js (hunk 28 above)
Comment (manumewari) [reply stacks/minstack.js]: added now

File: stacks/checkbrackets.js (hunk 31 above)
Comment (manumewari) [reply stacks/checkbrackets.js:3]: will make another one with 3 type brackets

File: stacks/checkbrackets.test.js (hunk 32 above)
Comment (manumewari) [reply stacks/checkbrackets.test.js:9]: removed duplicate one
